File: php/src/VedaPipeline.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Pipeline for batching multiple commands into a single round-trip.
 */
class VedaPipeline
{
    private VedaClient $client;

    /** @var list<array{type: string, sql: string}> */
    private array $commands = [];

    /** @var list<VedaResult> */
    private array $results = [];

    private bool $executed = false;

    public function __construct(VedaClient $client)
    {
        $this->client = $client;
    }

    /**
     * Add a query command to the pipeline.
     */
    public function query(string $sql): self
    {
        $this->commands[] = ['type' => 'query', 'sql' => $sql];
        return $this;
    }

    /**
     * Add an execute command to the pipeline.
     */
    public function execute(string $sql): self
    {
        $this->commands[] = ['type' => 'execute', 'sql' => $sql];
        return $this;
    }

    /**
     * Add an insert command to the pipeline.
     *
     * @param array<string, mixed> $data
     */
    public function insert(string $table, array $data): self
    {
        $cols = implode(', ', array_keys($data));
        $vals = implode(', ', array_map(
            fn(mixed $v) => $this->formatValue($v),
            array_values($data),
        ));
        $sql = "INSERT INTO {$table} ({$cols}) VALUES ({$vals});";
        $this->commands[] = ['type' => 'execute', 'sql' => $sql];
        return $this;
    }

    /**
     * Add an update command to the pipeline.
     *
     * @param array<string, mixed> $set
     */
    public function update(string $table, array $set, ?string $where = null): self
    {
        $setClauses = [];
        foreach ($set as $key => $value) {
            $setClauses[] = "{$key} = " . $this->formatValue($value);
        }
        $sql = "UPDATE {$table} SET " . implode(', ', $setClauses);
        if ($where !== null) {
            $sql .= " WHERE {$where}";
        }
        $this->commands[] = ['type' => 'execute', 'sql' => $sql . ';'];
        return $this;
    }

    /**
     * Add a delete command to the pipeline.
     */
    public function delete(string $table, ?string $where = null): self
    {
        $sql = "DELETE FROM {$table}";
        if ($where !== null) {
            $sql .= " WHERE {$where}";
        }
        $this->commands[] = ['type' => 'execute', 'sql' => $sql . ';'];
        return $this;
    }

    /**
     * Add a raw SQL command.
     */
    public function raw(string $sql): self
    {
        $this->commands[] = ['type' => 'raw', 'sql' => $sql];
        return $this;
    }

    /**
     * Get the number of queued commands.
     */
    public function count(): int
    {
        return count($this->commands);
    }

    /**
     * Check if the pipeline is empty.
     */
    public function isEmpty(): bool
    {
        return empty($this->commands);
    }

    /**
     * Clear all queued commands.
     */
    public function clear(): self
    {
        $this->commands = [];
        $this->results  = [];
        $this->executed = false;
        return $this;
    }

    /**
     * Execute all queued commands and return results.
     *
     * @return list<VedaResult>
     */
    public function run(): array
    {
        if ($this->executed) {
            throw new VedaException('Pipeline has already been executed');
        }

        if (empty($this->commands)) {
            return [];
        }

        $this->executed = true;

        // Use PIPELINE command if supported, otherwise execute sequentially
        $batchSql = 'PIPELINE ' . json_encode(array_map(
            fn(array $cmd) => $cmd['sql'],
            $this->commands,
        ));

        try {
            $result = $this->client->query($batchSql);
            $this->results = $this->parseBatchResult($result);
        } catch (QueryException $e) {
            // Fallback to sequential execution
            $this->results = $this->executeSequential();
        }

        return $this->results;
    }

    /**
     * Execute commands sequentially as fallback.
     *
     * @return list<VedaResult>
     */
    private function executeSequential(): array
    {
        $results = [];
        foreach ($this->commands as $cmd) {
            try {
                $results[] = $this->client->query($cmd['sql']);
            } catch (\Throwable $e) {
                $results[] = new VedaResult(
                    null,
                    null,
                    0,
                    null,
                    ['error' => $e->getMessage()],
                );
            }
        }
        return $results;
    }

    /**
     * Parse a batch result into individual VedaResult objects.
     *
     * @return list<VedaResult>
     */
    private function parseBatchResult(VedaResult $result): array
    {
        // Server returns rows with each result as JSON
        $results = [];
        if ($result->rows !== null) {
            foreach ($result->rows as $row) {
                $raw = $row[0] ?? '{}';
                try {
                    $results[] = VedaResult::parse(is_string($raw) ? $raw : json_encode($raw));
                } catch (\Throwable $e) {
                    $results[] = new VedaResult(
                        null, null, 0, null, ['error' => $e->getMessage()],
                    );
                }
            }
        }

        // If no rows, return single result
        if (empty($results)) {
            $results[] = $result;
        }

        return $results;
    }

    /**
     * Get results after run().
     *
     * @return list<VedaResult>
     */
    public function getResults(): array
    {
        if (!$this->executed) {
            throw new VedaException('Pipeline has not been executed. Call run() first.');
        }
        return $this->results;
    }

    /**
     * Get a result at a specific index.
     */
    public function getResult(int $index): ?VedaResult
    {
        if (!$this->executed) {
            throw new VedaException('Pipeline has not been executed');
        }
        return $this->results[$index] ?? null;
    }

    private function formatValue(mixed $value): string
    {
        if ($value === null) return 'NULL';
        if (is_bool($value)) return $value ? 'TRUE' : 'FALSE';
        if (is_string($value)) return "'" . str_replace("'", "''", $value) . "'";
        return (string) $value;
    }
}

File: dotnet/VedaPipeline.cs
using System;

namespace VedaDB;

/// <summary>
/// A single command in a pipeline.
/// </summary>
public class VedaPipelineCommand
{
    /// <summary>
    /// SQL command text.
    /// </summary>
    public string Sql { get; set; } = "";

    /// <summary>
    /// Parameters for the command.
    /// </summary>
    public object[] Parameters { get; set; } = Array.Empty<object>();

    /// <summary>
    /// Whether this is a write operation.
    /// </summary>
    public bool IsWrite { get; set; }
}

/// <summary>
/// Pipeline execution result.
/// </summary>
public class VedaPipelineResult
{
    /// <summary>
    /// Results for each command in order.
    /// </summary>
    public List<VedaResult> Results { get; set; } = new();

    /// <summary>
    /// Any errors that occurred.
    /// </summary>
    public List<(int Index, string Error)> Errors { get; set; } = new();

    /// <summary>
    /// Total number of commands.
    /// </summary>
    public int CommandCount { get; set; }

    /// <summary>
    /// Number of successful commands.
    /// </summary>
    public int SuccessCount { get; set; }

    /// <summary>
    /// Whether all commands succeeded.
    /// </summary>
    public bool AllSucceeded => Errors.Count == 0;

    /// <summary>
    /// Total duration of the pipeline.
    /// </summary>
    public TimeSpan Duration { get; set; }
}

/// <summary>
/// Command pipeline for batching multiple VedaDB operations.
/// Reduces round-trips by sending commands together.
/// </summary>
public class VedaPipeline
{
    private readonly VedaClient _client;
    private readonly List<VedaPipelineCommand> _commands = new();
    private bool _isAtomic;

    /// <summary>
    /// Number of commands in the pipeline.
    /// </summary>
    public int Count => _commands.Count;

    /// <summary>
    /// Whether to execute as a transaction (all or nothing).
    /// </summary>
    public bool IsAtomic => _isAtomic;

    /// <summary>
    /// Create a pipeline.
    /// </summary>
    public VedaPipeline(VedaClient client)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
    }

    /// <summary>
    /// Add a query command to the pipeline.
    /// </summary>
    public VedaPipeline AddQuery(string sql, params object[] parameters)
    {
        _commands.Add(new VedaPipelineCommand
        {
            Sql = sql,
            Parameters = parameters,
            IsWrite = false
        });
        return this;
    }

    /// <summary>
    /// Add a write command (INSERT/UPDATE/DELETE) to the pipeline.
    /// </summary>
    public VedaPipeline AddExecute(string sql, params object[] parameters)
    {
        _commands.Add(new VedaPipelineCommand
        {
            Sql = sql,
            Parameters = parameters,
            IsWrite = true
        });
        return this;
    }

    /// <summary>
    /// Add an INSERT command.
    /// </summary>
    public VedaPipeline AddInsert(string table, Dictionary<string, object?> data)
    {
        var cols = string.Join(", ", data.Keys);
        var vals = string.Join(", ", data.Values.Select(FormatValue));
        return AddExecute($"INSERT INTO {table} ({cols}) VALUES ({vals});");
    }

    /// <summary>
    /// Add an UPDATE command.
    /// </summary>
    public VedaPipeline AddUpdate(string table, Dictionary<string, object?> set, string? where = null)
    {
        var setClause = string.Join(", ", set.Select(kv => $"{kv.Key} = {FormatValue(kv.Value)}"));
        var sql = $"UPDATE {table} SET {setClause}";
        if (!string.IsNullOrEmpty(where)) sql += $" WHERE {where}";
        return AddExecute(sql + ";");
    }

    /// <summary>
    /// Add a DELETE command.
    /// </summary>
    public VedaPipeline AddDelete(string table, string? where = null)
    {
        var sql = $"DELETE FROM {table}";
        if (!string.IsNullOrEmpty(where)) sql += $" WHERE {where}";
        return AddExecute(sql + ";");
    }

    /// <summary>
    /// Mark this pipeline as atomic (transactional).
    /// </summary>
    public VedaPipeline Atomic()
    {
        _isAtomic = true;
        return this;
    }

    /// <summary>
    /// Clear all commands from the pipeline.
    /// </summary>
    public void Clear() => _commands.Clear();

    /// <summary>
    /// Execute all commands in the pipeline.
    /// </summary>
    public async Task<VedaPipelineResult> ExecuteAsync(CancellationToken ct = default)
    {
        if (_commands.Count == 0)
            return new VedaPipelineResult();

        var sw = System.Diagnostics.Stopwatch.StartNew();
        var result = new VedaPipelineResult { CommandCount = _commands.Count };

        try
        {
            if (_isAtomic)
            {
                await _client.ExecuteAsync("BEGIN");
            }

            foreach (var (cmd, index) in _commands.Select((c, i) => (c, i)))
            {
                try
                {
                    var sql = cmd.Sql;
                    if (cmd.Parameters.Length > 0)
                        sql = BindParameters(sql, cmd.Parameters);

                    VedaResult cmdResult;
                    if (cmd.IsWrite)
                    {
                        var rowsAffected = await _client.ExecuteAsync(sql);
                        cmdResult = new VedaResult { Message = rowsAffected.ToString(), RowCount = rowsAffected };
                    }
                    else
                    {
                        cmdResult = await _client.QueryAsync(sql);
                    }

                    result.Results.Add(cmdResult);
                    result.SuccessCount++;
                }
                catch (Exception ex)
                {
                    result.Errors.Add((index, ex.Message));

                    if (_isAtomic)
                    {
                        await _client.ExecuteAsync("ROLLBACK");
                        sw.Stop();
                        result.Duration = sw.Elapsed;
                        return result;
                    }
                }
            }

            if (_isAtomic && result.AllSucceeded)
            {
                await _client.ExecuteAsync("COMMIT");
            }

            sw.Stop();
            result.Duration = sw.Elapsed;

            VedaMetrics.Increment("vedadb_pipeline_executions", 1,
                new() { { "atomic", _isAtomic.ToString() } });
            VedaMetrics.Histogram("vedadb_pipeline_duration", sw.Elapsed.TotalSeconds);
            VedaMetrics.Gauge("vedadb_pipeline_commands", _commands.Count);

            return result;
        }
        catch
        {
            if (_isAtomic)
            {
                try { await _client.ExecuteAsync("ROLLBACK"); } catch { /* Best effort */ }
            }
            throw;
        }
        finally
        {
            _commands.Clear();
        }
    }

    /// <summary>
    /// Execute and return results as a list.
    /// </summary>
    public async Task<List<VedaResult>> ExecuteAndGetResultsAsync(CancellationToken ct = default)
    {
        var pipelineResult = await ExecuteAsync(ct);
        return pipelineResult.Results;
    }

    private static string BindParameters(string sql, object[] parameters)
    {
        var result = sql;
        for (int i = 0; i < parameters.Length; i++)
        {
            result = result.Replace($"@{i}", FormatValue(parameters[i]));
        }
        return result;
    }

    private static string FormatValue(object? value) => value switch
    {
        null => "NULL",
        string s => $"'{s.Replace("'", "''")}'",
        bool b => b ? "TRUE" : "FALSE",
        DateTime dt => $"'{dt:yyyy-MM-dd HH:mm:ss}'",
        _ => value.ToString() ?? "NULL"
    };
}
